keps/sig-network/1024-nodelocal-cache-dns/README.md

```diff
@@ -119,6 +131,98 @@ We still need some component to dynamically determine when to use node-local-dns
+| Single Service        | Unbound | Yes     | 3287 |
+|                       |         |         |      |
+| Single NXDomain       | CoreDNS | Yes     | 816  |
+| Single NXDomain       | Unbound | Yes     | 3136 |
+|                       |         |         |      |
+| Multiple services(20) | CoreDNS | No      | 859  |
+| Multiple services(20) | Unbound | No      | 1463 |
+|                       |         |         |      |
+| External queries      | CoreDNS | No      | 180  |
+| External queries      | Unbound | No      | 108  |
+|                       |         |         |      |
+| Single Service        | CoreDNS | No      | 818  |
+| Single Service        | Unbound | No      | 2992 |
+|                       |         |         |      |
+| Single NXDomain       | CoreDNS | No      | 827  |
+| Single NXDomain       | Unbound | No      | 2986 |
+
+
+Peak memory usage was ~20 Mi for both Unbound and CoreDNS.
+
+For the single service and single NXDomain query, Unbound still had cache hits since caching could not be completely disabled.
+
+CoreDNS QPS was twice as much as Unbound for external queries. They were mostly unique hostnames from this file - [ftp://ftp.nominum.com/pub/nominum/dnsperf/data/queryfile-example-current.gz](ftp://ftp.nominum.com/pub/nominum/dnsperf/data/queryfile-example-current.gz)
+
+When multiple cluster services were queried with cache misses, Unbound was better(1463 vs 859), but not by a large factor.
+
+Unbound performs much better when all requests are cache hits.
+
+CoreDNS will be the local cache agent in the first release, after considering these reasons:
+
+*  Better QPS numbers for external hostname queries
+*  Single process, no need for a separate nanny process
+*  Prometheus metrics already available, also we can get per zone stats. Unbound gives consolidated stats.
+*  Easier to make changes to the source code
+
+ It is possible to run any program as caching agent by modifying the daemonset and configmap spec. Publishing an image with Unbound DNS can be added as a follow up.
+
+Based on the prototype/test results, these are the recommended defaults: 
+CPU request: 50m
+Memory Limit : 25m  
+
+CPU request can be dropped to a smaller value if QPS needs are lower.
+
+#### Metrics
+
+Per-zone metrics will be available via the metrics/prometheus plugin in CoreDNS.
+
+
+### Risks and Mitigations
+
+* The proposed HA solution will not work in IPVS mode of kube-proxy. This is because skipping IPVS translation rules is not possible using an iptables NOTRACK rule. So, if the service IP is used by pods for DNS resolution, requests will always hit the IPVS load-balancing rules and reach the kube-dns endpoints.
+One way to get the desired behavior is to change the selectors for the kube-dns service, so that we set empty endpoints when we want the node-local-dns pods to be used. However, this approach has not been tested yet. Also, this would change it across the board, instead of for a single node as the need may be.
```

Comment:
I think this can be removed. Not sure setting empty endpoints will work in IPVS.

Reply:
removed